lib/pure/asynchttpserver.nim

```diff
@@ -193,13 +214,30 @@ proc parseProtocol(protocol: string): tuple[orig: string, major, minor: int] =
 proc sendStatus(client: AsyncSocket, status: string): Future[void] =
   client.send("HTTP/1.1 " & status & "\c\L\c\L")
 
+when (NimMajor, NimMinor) >= (1, 1):
+  iterator bodyStream*(
+    request: Request,
+    chunkSize: int = 8*1024): (int, Future[string]) =
+    ## The chunkSize parameter is optional and default value is 8*1024 bytes.
+    ##
+    ## This iterator return a tuple with the length of the data that was read
+    ## and a future.
+    var remainder = request.contentLength
+    while remainder > 0:
+      let readSize = min(remainder, chunkSize)
+      let data = request.client.recv(readSize)
+      if data.failed:
+        raise newException(ValueError, "Error reading POST data from client.")
+      yield (readSize, data)
+      remainder -= readSize
```

Comment:
This iterator is quite broken and makes me wonder whether you've tested this at all:

* `recv` reads **up to** `readSize`, it doesn't guarantee you'll get that much data, and a lot of the time you won't.

* you're checking whether the `data` future failed, without `await`ing it. This code path will almost never trigger.

Reply:
You can test with the example and you will see that it works. Try uploading a 900 MB file!

Reply:
> This iterator is quite broken and makes me wonder whether you've tested this at all:
> 
>     * `recv` reads **up to** `readSize`, it doesn't guarantee you'll get that much data, and a lot of the time you won't.
> 
>     * you're checking whether the `data` future failed, without `await`ing it. This code path will almost never trigger.
From the nim documentation (https://nim-lang.org/0.16.0/asyncdispatch.html)
```nim
var future = sock.recv(100)
yield future
if future.failed:
  # Handle exception
```
I can't use the await in the iterators!
https://github.com/nim-lang/Nim/issues/3885